doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+# Namespace and Name of Keys
+
+## Problem
+
+A `Key` in Elektra is identified by its name, which consists of a namespace and a number of name parts.
+There are two common representations for the name: escaped and unescaped.
+
+The unescaped form is essentially a single namespace byte plus and arbitrary-length sequence of arbitrary bytes, in which a zero byte separates parts.
+The escaped name is a more complicated representation as a zero-terminated string that maps 1:1 onto unescaped names.
+More details can be found in [the relevant docs](/doc/KEYNAMES.md).
+
+There is a conflict between these two forms in terms of API convenience and efficient execution.
+Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
+When it comes to execution, however, the unescaped name often makes things more efficient.
+For example, order comparisons between unescaped names are much simpler.
+
+The question now is, which representations should be used by `libelektra-core` and how.
+
+## Constraints
+
+- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be fast.
+
+## Assumptions
+
+- In most cases the escaped name is used for developer convenience not, because of actual requirements.
+- The most common requirement for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
```

Comment:
Also, debugging. Deciphering the unescaped name in debuggers is horrible. But this could be solved by providing [custom pretty-printers](https://undo.io/resources/gdb-watchpoint/here-quick-way-pretty-print-structures-gdb/) for gdb and other debuggers.

Reply:
In GDB (and probably others) you can do:

```
p *key->ukey@key->keyUSize
```

This prints `key->ukey` as a fixed-length string of length `key->keyUSize`, e.g.:

```
$1 = "\006\000abc"
```

Had to google that syntax myself, but I was 99.9% sure it existed.

Reply:
How is that "rationale", it hard explains the decision and "easier debugging" really shouldn't affect an API related decision. I'll add it to the notes though.